client/src/http.rs

```diff
@@ -1,35 +1,337 @@
+/// all methods. Only [`RequestBuilder::new()`] is the required one.
 pub trait RequestBuilder {
-    /// Constructs a new builder with provided method and URL
-    #[must_use]
-    fn new<U>(method: Method, url: U) -> Self
-    where
-        U: AsRef<str>;
-
-    /// Sets request's body in bytes
+    /// Entrypoint - create a new builder with specified method and URL.
     #[must_use]
-    fn bytes(self, data: Vec<u8>) -> Self;
+    fn new(method: Method, url: impl AsRef<str>) -> Self;
 
-    /// Sets request's query params
+    /// Add multiple query params at once. Uses [`RequestBuilder::param()`] for each param.
     #[must_use]
-    fn params<P, K, V>(self, params: P) -> Self
+    fn params<P, K, V>(mut self, params: P) -> Self
     where
         P: IntoIterator,
         P::Item: Borrow<(K, V)>,
         K: AsRef<str>,
+        V: ToString,
+        Self: Sized,
+    {
+        for pair in params {
+            let (k, v) = pair.borrow();
+            self = self.param(k, v.to_string());
+        }
+        self
+    }
+
+    /// Add a single query param
+    #[must_use]
+    fn param<K, V>(self, key: K, value: V) -> Self
+    where
+        K: AsRef<str>,
+        V: ToString;
+
+    /// Add multiple headers at once. Uses [`RequestBuilder::header()`] for each param.
+    #[must_use]
+    fn headers<H, N, V>(mut self, headers: H) -> Self
+    where
+        H: IntoIterator,
+        H::Item: Borrow<(N, V)>,
+        N: AsRef<str>,
+        V: ToString,
+        Self: Sized,
+    {
+        for pair in headers {
+            let (k, v) = pair.borrow();
+            self = self.header(k, v.to_string());
+        }
+        self
+    }
+
+    /// Add a single header
+    #[must_use]
+    fn header<N, V>(self, name: N, value: V) -> Self
+    where
+        N: AsRef<str>,
         V: ToString;
 
-    /// Sets request's headers
+    /// Set request's binary body
     #[must_use]
-    fn headers(self, headers: Headers) -> Self;
+    fn body(self, data: Vec<u8>) -> Self;
+}
+
+/// Generalization of `WebSocket` client's functionality
+pub mod ws {
+    use super::{eyre, RequestBuilder, Result};
+
+    /// `WebSocket` connection flow stages.
+    ///
+    /// Flow consists of the following:
+    ///
+    /// 1. **Init stage** - establish `WebSocket` connection with Iroha
+    /// 2. **Handshake stage** - send a "subscription" message to Iroha and ensure that the next message from Iroha
+    ///     is a "subscription accepted" message
+    /// 3. **Events stage** - wait for messages from Iroha. For each message, decode *some event* from it
+    ///     and send back *some "received"* mesage
+    ///
+    ///
+    ///
+    /// This module has a set of abstraction to extract pure data logic from transportation logic. Following sections
+    /// describe how to use this module from both **flow implemention** (data side) and
+    /// **transport implementation** sides.
```

Comment:
I don't know if this much abstraction is necessary. I guess the following works just as well:

```rust
impl Client {
    fn subscribe(&self, filter: EventFilter) -> Result<WSClient> {
        //
    }
}

struct WSClient {
    //
    stream: Stream,
    //
}

impl WSClient {
    fn receive(&self) -> Result<Event> {
        //
    }
}
```

Reply:
But how in this case it is possible to use your own HTTP/WS client?

Reply:
Frankly, I still don't see the whole picture.
However, since this PR seems to have subsequent one, I'd like to interfere with your concept as little as possible.
I think it is not too late to make a suggestion like this once all requirements have been met